Convert planetary system data to yaml + general planetary system overhaul

AaronGullickson opened the pull request:
The primary goal of this PR is to convert the current xml encoding of planetary systems into a yaml format that is read in through Jackson. In the process of doing so, this PR cleans up a lot of the planetary system code, organizes the data in an easier fashion for editing, and adds a few new features. In addition to changing the overall format from xml to yaml, the PR also changes the file structure so that each planetary system's data is held in a separate file within `data/universe/planetary_systems` in two separate subfolders labeled `canon_systems` and `connector_systems` (the latter being used for fake systems we use to connect far-flung periphery locations and the Clan Homeworlds). Thus, the large number of files being touched here is very misleading, about the scale of the actual changes (which are still substantial).

## Rationale

The rationale for changing to yaml is twofold. First, when it is necessary to edit planetary system data by hand the current set up is unwieldy. Base planet data is in a single `systems.xml` file and planetary events like population, faction change, and SICS code are in a separate `system_events.xml`. Finding the correct planet is thus a chore and these files are huge making it difficult to edit, especially for less technically inclined volunteers. In fact, you can't even edit the files in Intellij because they are too big. The new set up puts all the data for a single planetary system in a separate file in a format that is easier to read and more intuitive for most people. 

However, our ultimate goal is to avoid as much hand editing as possible. In that sense, this PR is a precursor to some other projects I have in the works to create backend web-based [Shiny](https://www.shinyapps.io/) apps that will make it possible to load and edit planetary system files via a browser. The yaml format will be much easier to work with for those projects than the current xml. 

## Data Conversion Process

I used R scripts in [this](https://github.com/AaronGullickson/convert_planetary_systems) repository to convert from the old xml code to the new yaml code. In doing so I made the following improvements to the data:

- Added source information where it was non-canon using a `source:` `value:` pairing for all entries. See below for details and examples. 
- Cleaned up the underlying data for connector systems, which wasn't working properly. We had no icons for the planets in these systems and satellites were coded using the wrong XML syntax so they all ended up null. Connector systems all look great now.
- In the previous XML transition, we included some canon population data, but we never properly indicated the source for this data in the XML data. So, I added that source information to the new yaml data.
- Where we are using enums, I replaced the raw text data with the correct enum value. For example, in atmosphere I changed `Toxic (Poisonous)` to `TOXICPOISON`.

We also wanted to better capture the source information for each entry, which was entered into XML as an attribute. We handle this by breaking every entry into a `source:` `value:` pair. For example here is the entry for Terra's gravity:

```yaml
gravity:
  source: canon
  value: 1.0
```

Ideally, as we enter stuff in the future, we can be more specific than just "canon" but that is what we have for most canon entries right now.  For cases, where do not have a canon entry, we just leave out the source part entirely, for example:

```yaml
pressure:
  value: HIGH
```

This makes the files a little longer and less easy to read, but is critically important for ensuring proper sourcing of all data whenever possible. I create a `SourceableValue` class in java discussed below to capture this data structure from the yaml.

## Code Improvements

To read the new data structure in smoothly using Jackson, I had to make some changes to the planetary system data that have been long overdue. These changes are:

- Create a `StarType` class to track information about a given star, including calculating internally things like time to jump point and solar charging time.  To do this I moved a bunch of stuff out of `StarUtil`. Most of the stuff left in `StarUtil` is currently unused but I decided to leave it rather than try to get it all sorted for this PR.  If the reviewers would prefer I do that now, I can.
- Create proper enums for HPG rating and planetary type. The old HPG rating code relied on EquipmentType integers which was ugly and not a great fit because HPGs only go from A to D. Plus equipment type integers are reverse ordered. Planetary type was just a string. 
- Created a `LandMass` class to record information about each landmass on a planet.
- Cleaned out a **lot** of unused out of date classes and methods. I was pretty ruthless here. I think we have a tendency to leave this stuff in which then makes the code harder to read, especially for newbies to the project. Its all saved in git if we need it again.
- Made quite a few changes to the loading procedure to simplify it. One nice benefit of reading in individual planet files is that we can catch IO errors and not break the whole thing instead  just lose that planet with the io error reported to the log.
- Added the `SourceableValue` class which I disuss below.

## SourceableValue

Probably one of the biggest changes is the use of the [SourceableValue](https://github.com/AaronGullickson/mekhq/blob/planetary-yaml/MekHQ/src/mekhq/campaign/universe/SourceableValue.java) class. This is a generic class which records both the source (as a string) and value (as a generic type) for any possible entry. I then use this to record any base planetary data or event that we consider sourceable, which is almost everything. the `source` value is null for non-canon entries. 

Currently, we are not doing anything with this information in MekHQ, but a later PR goal will be to provide this information to users in some way through the `PlanetViewPanel` (I am thinking maybe as a mouseover popup). However, this information is critical in our yaml data for external backend editor development. So, I needed some way to load the data structure using Jackson and this class works very smoothly.

This class does make it a little harder to get desired values out of planetary systems. To handle this I created a set of `getSourced*` functions to retrieve sourceableValues and then rewrite all the getter methods to call these functions to get values. 

## Custom User Data

Most of the stuff in this PR is backend and will not affect users, but there will be one nice new user feature. If users have a custom data directory defined in the `mmconf` file, they can include a `data/universe/planetary_systems` directory. Any yaml file they drop in there will be read as a custom planetary system that will be loaded when MekHQ launches. 

We did have some relic code for reading in custom planetary system events. However, since the last big XML makeover, we have not provided users a way to add or edit  events, so there shouldn't actually be anything there. A future PR will build in a custom events editor that will allow players to add custom events to a given planetary system that will get saved in the campaign file. So if players take over a planet for the Dracs, they can record that, or if they totally nuke a planet from orbit, they can also record that. But for now, I just removed all the legacy code for custom planetary events since there is no more XML code to load/save. If, for some reason, players have this in their campaign files, it will be ignored and not cause any errors in loading. 

The last possibility is that some users may have created entire AUs by either writing their own systems.xml or adding in other xml files. These will no longer work. I would like to at some point offer up a Shiny app for converting old xml files like this to the new yaml format, but my guess is there are relatively few players in this category.

## Refactoring

It might be a good idea to refactor all of the planetary stuff in the `universe` directory into a `planetary_system` subfolder. I did not do that here because I thought it might make it harder for reviewers to review the work and might cause a lot of chaos with other PRs people are working on, but if that is a desire of reviewers I can do so. 

**EDIT**: One more thing I forgot to mention is that I would like to add something to the gradle packaging scripts that zips up the planet files for distribution, just like mekfiles in MegaMek. They can be read directly from a zip file. @rjhancock is working on some gradle code to do that because I don't have a lot of gradle experience.

AaronGullickson commented:
This PR is now ready for review, although we should wait to merge until we have added gradle code to zip up the planetary system files for distribution.

AaronGullickson commented:
It does appear that the large number of files here is kind of breaking the GH review ability.

AaronGullickson commented:
Ok, I backed out the actual data into a separate branch off this PR that can be viewed [here](https://github.com/AaronGullickson/mekhq/tree/actual-yaml-data/MekHQ/data/universe/planetary_systems). That should make it possible to review the code in this PR, but once that code is approved, **please do not merge the PR**. I will first merge the data back in so that we get it all merged in together.

rjhancock commented on MekHQ/src/mekhq/campaign/universe/StarType.java:
```diff
@@ -0,0 +1,422 @@
+            subtypeFormat = "%.1f";
+        }
+
+        if (null != luminosity && luminosity.equals(StarType.LUM_VI)) {
+            // subdwarfs
+            return "sd" + getSpectralClassName(spectralClass) + String.format(subtypeFormat, subtypeValue / 100.0);
+        } else if (null != luminosity && luminosity.equals(StarType.LUM_VI_PLUS)) {
+            // extreme subdwarfs
+            return "esd" + getSpectralClassName(spectralClass) + String.format(subtypeFormat, subtypeValue / 100.0);
+        } else if (null != luminosity && luminosity.equals(StarType.LUM_VII)) {
+            // white dwarfs
+            return String.format(Locale.ROOT, "D" + subtypeFormat, subtypeValue / 100.0);
+        } else {
+            // main class
+            return String.format(Locale.ROOT, "%s" + subtypeFormat + "%s",
+                getSpectralClassName(spectralClass),
+                subtypeValue / 100.0, (null != luminosity ? luminosity : StarType.LUM_V));
+        }
```
This can be made more legible with an initial escape with a `null != luminosity` check and return whatever the default is then run a switch for the other 3 followed by the default WITH luminosity.

rjhancock commented on MekHQ/src/mekhq/campaign/universe/enums/HPGRating.java:
```diff
@@ -0,0 +1,57 @@
+        return name;
+    }
+
+    public static HPGRating parseHPGRating(String val) {
+        try {
+            return HPGRating.valueOf(val.toUpperCase());
+        } catch (Exception ex) {
+            Sentry.captureException(ex);
+            logger.error("Couldn't find a HPG rating level matching " + val.toUpperCase(), ex);
+            return X;
+        }
```
The logger will auto-report to Sentry if enabled. No need to include and call it explicitly.

There should also be a method on Logger that takes the EX first followed by a string and a varargs.

rjhancock commented on MekHQ/src/mekhq/campaign/universe/enums/HiringHallLevel.java:
```diff
@@ -18,8 +18,12 @@
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
 import io.sentry.Sentry;
```
No need to include. There should be no place within the code that needs this outside of Main and the Logger.

AaronGullickson replied:
Yeah, whoops that was a relic from when I using a custom deserializer because I didn't realize Jackson would get it right if you fed in the raw enum value. For some reason Intellij is not showing me unused imports, which I thought it used to do.

rjhancock replied:
I think it's a setting you have to change. This stuff is always flagged on my VSCode but I can't rely on it to do MM work on it. Project is big enough there's a 50/50 chance it'll work right.

rjhancock commented on MekHQ/src/mekhq/campaign/universe/enums/PlanetaryType.java:
```diff
@@ -1,5 +1,7 @@
 /*
- * Copyright (c) 2019-2022 - The MegaMek Team. All Rights Reserved.
+ * PlanetaryType.java
+ *
+ * Copyright (C) 2025 MegaMek team
```
Can update the year but can't change the copyright line.

rjhancock reviewed (commented):
Sanity check complete. With the exception of a copyright and the library versions, most are nitpick things to make it easier for other (and newer) developers to work on the code if needed and just some questions about some method removals.

Overall I like the way the code is going and glad to see large swaths of old code get removed and updated.

rjhancock commented:
Make a PR for the adding of the new data itself as that can be incorporated without much review and should be done before this is merged.

AaronGullickson commented on MekHQ/src/mekhq/campaign/universe/Atmosphere.java:
```diff
@@ -28,20 +32,6 @@ public enum Atmosphere {
     BREATHABLE("Breathable");
 
-    // For old life form data
-    public static Atmosphere parseAtmosphere(String val) {
-        return switch (val.toLowerCase()) {
-            case "tainted (poisonous)" -> TAINTEDPOISON;
-            case "tainted (caustic)" -> TAINTEDCAUSTIC;
-            case "tainted (flammable)" -> TAINTEDFLAME;
-            case "toxic (poisonous)" -> TOXICPOISON;
-            case "toxic (caustic)" -> TOXICCAUSTIC;
-            case "toxic (flammable)" -> TOXICFLAME;
-            case "breathable" -> BREATHABLE;
-            default -> NONE;
-        };
-    }
```
Just now. With Jackson, its simpler to read in enum variables if you use the actual enum values in the yaml data - otherwise you have to define a custom deserializer. So, I switched all the data in the yaml from the text description (e.g. `Toxic (Poisonous)`) to the enum value (e.g. `TOXICPOISON`). The XML adapter was the only thing to use this method so it wasn't needed anymore.

AaronGullickson commented on MekHQ/src/mekhq/campaign/universe/LifeForm.java:
```diff
@@ -18,40 +18,14 @@
     NONE("None"), MICROBE("Microbes"), PLANT("Plants"), INSECT("Insects"), FISH("Fish"),
     AMPHIBIAN("Amphibians"), REPTILE("Reptiles"), BIRD("Birds"), MAMMAL("Mammals");
 
-    // For old life form data
-    public static LifeForm parseLifeForm(String val) {
-        switch (val.toLowerCase()) {
-            case "microbe":
-            case "microbes":
-                return MICROBE;
-            case "plant":
-            case "plants":
-                return PLANT;
-            case "fish":
-                return FISH;
-            case "amphibian":
-            case "amphibians":
-                return AMPHIBIAN;
-            case "reptile":
-            case "reptiles":
-                return REPTILE;
-            case "bird":
-            case "birds":
-                return BIRD;
-            case "mammal":
-            case "mammals":
-                return MAMMAL;
-            case "insect":
-            case "insects":
-                return INSECT;
-            default: return NONE;
-        }
-    }
-
```
Same answer as above - with this PR.

AaronGullickson commented on MekHQ/src/mekhq/campaign/universe/Planet.java:
```diff
@@ -579,41 +442,18 @@ public PlanetaryEvent getCurrentEvent(LocalDate now) {
-            }
-        }
-        return result;
+    public String getName(LocalDate when) {
```
Would you like me to leave them in with deprecated tag?

rjhancock replied:
Unless we can be certain they are not being used elsewhere, please add them back in with the tag and document as such. After the next milestone we can work to remove them proper.

SJuliez replied:
I'd like to say that I have not been following such a rule. Intellij with MHQ (!) loaded tells me if a method is used and if it is not and there's no other reason to keep it (debug methods or classes) then I remove it. With all the changes I make I would be unable to remember what to revisit and remove at a later release. The chances for an unused method to suddenly become important are slim if not nonexistent. And, in any case, the code is never truly lost in the repo. Therefore, I'd like to stick with simply going forward with code changes if they seem good.

rjhancock replied:
> I'd like to say that I have not been following such a rule. Intellij with MHQ (!) loaded tells me if a method is used and if it is not and there's no other reason to keep it (debug methods or classes) then I remove it. 

I generally agree with this however there have been times where I've made changes in one program and had to adjust it in others. Although less likely to happen in MekHQ, it can happen that the IDE doesn't catch all instances.

> With all the changes I make I would be unable to remember what to revisit and remove at a later release. 

Hence the point of marking them as Deprecated. Something we can go back and search for to remove. Once a milestone is announced, we go back and remove all items marked as such or with another comment talking about "Remove after next milestone" to keep the code clean.

> The chances for an unused method to suddenly become important are slim if not nonexistent. 

I've had methods in the entire MM suite and other projects where a method would be marked as not in use and doing a global search had it pop up in 30 places. The problem with this code base is that it is so large with multiple people working on it, one PR could remove a method that another PR may depend upon and a replacement isn't available. One developer decided it was not important and removed it, the other deemed it of value. Yes it's in the history and can be pulled back in, but marking it Deprecated gives time to determine if it is being used in some other branch.

> And, in any case, the code is never truly lost in the repo. Therefore, I'd like to stick with simply going forward with code changes if they seem good.

I'm not saying to not do this, I'm only asking that methods be marked deprecated for one milestone cycle BEFORE we remove them.

AaronGullickson replied:
For now, I put it back in with the deprecated tag.

AaronGullickson commented on MekHQ/src/mekhq/campaign/universe/Planet.java:
```diff
@@ -738,10 +612,14 @@ public boolean isHiringHall(LocalDate when) {
      */
     public HiringHallLevel getHiringHallLevel(LocalDate when) {
-        HiringHallLevel staticHall = getEventData(when, hiringHall, e -> e.hiringHall);
+        HiringHallLevel staticHall = (null == getSourcedHiringHallLevel(when)) ?
+            HiringHallLevel.NONE : getSourcedHiringHallLevel(when).getValue();
```
All of the `getSourcedX` method returns the actual `SourceableValue` object so you can't really have it return an enum. I do agree all the null checks are a bit tedious, but I don't have another workaround. I think it has to be checked somewhere.

rjhancock replied:
Moving the null checks to the `getSourcedX` methods would remove the extra null checks as well as the double calls and processing.

Not sure how much extra work that would be, but it would simplify the call sites.

AaronGullickson commented on MekHQ/src/mekhq/campaign/universe/PlanetarySystem.java:
```diff
@@ -746,4 +505,41 @@ public String getAcademiesForSystem(List<Academy> filteredAcademies) {
+    public static class PlanetarySystemPostLoader extends StdConverter<PlanetarySystem, PlanetarySystem> {
+
+        @Override
+        public PlanetarySystem convert(PlanetarySystem ps) {
+            if (null == ps.id) {
+                ps.id = ps.name;
+            }
+
+            // fill up planets
+            ps.planets = new TreeMap<>();
+            if (null != ps.planetList) {
+                for (Planet p : ps.planetList) {
+                    p.setParentSystem(ps);
+                    if (!ps.planets.containsKey(p.getSystemPosition())) {
+                        ps.planets.put(p.getSystemPosition(), p);
+                    }
+                }
+                ps.planetList.clear();
+            }
+            ps.planetList = null;
+            // Fill up events
+            ps.events = new TreeMap<>();
+            if (null != ps.eventList) {
+                for (PlanetarySystemEvent event : ps.eventList) {
+                    if ((null != event) && (null != event.date)) {
+                        ps.events.put(event.date, event);
+                    }
+                }
+                ps.eventList.clear();
+            }
+            ps.eventList = null;
+
+            return ps;
+        }
```
Definitely, can do.

AaronGullickson commented:
> Make a PR for the adding of the new data itself as that can be incorporated without much review and should be done before this is merged.

Wouldn't it be easier to just add it here before merging the PR?

AaronGullickson commented:
Ok, I think I addressed all of your comments with code changes where I could. Wasn't sure if you were asking me to put those functions back in there as deprecated.

rjhancock commented:
Unless you're absolutely certain they are not used elsewhere in the project, please put those methods back in and marked @Depricated with comments. Either will work for me.

Trying to make sure we get in the habit of before we remove a method, deprecate it, move the code off of it, then remove it as part of the first set of PRs after a milestone release. Try to keep the code clean and updated kind of thing and reduce the tech debt.

rjhancock commented:
> > Make a PR for the adding of the new data itself as that can be incorporated without much review and should be done before this is merged.
> 
> Wouldn't it be easier to just add it here before merging the PR?

It can go either way. My thought process is given the amount of data, do one PR for that then one for this. I wouldn't hold this up if you re-committed the data to this PR for the final merge.

SJuliez commented:
It is possible to test a yaml node in Jackson if it is a value node (pressure: HIGH) or has children (pressure: // value: HIGH). This way both variants could be parsed. I feel this would enhance the yaml files considerably.

I did precisely that at various points for parsing MM scenario files.

SJuliez reviewed (approved):
Code looks good. If doable, I feel that enabling it to read values inline instead of with a mandatory value: tag would improve the yaml quality a lot.

AaronGullickson commented on MekHQ/src/mekhq/campaign/universe/Planet.java:
```diff
@@ -579,41 +442,18 @@ public PlanetaryEvent getCurrentEvent(LocalDate now) {
         return getEventData(when, shortName, e -> e.shortName);
     }
 
-    public List<String> getNames() {
-        List<String> names = new ArrayList<>();
-
-        for (PlanetaryEvent p : events.values()) {
-            names.add(p.name);
-        }
-
-        return names;
-    }
-
```
I tried this but with the addition of the `SourceableValue`, you can no longer just get a name out of the planetary event `p` as it was being done on line 611.

AaronGullickson commented:
> It is possible to test a yaml node in Jackson if it is a value node (pressure: HIGH) or has children (pressure: // value: HIGH). This way both variants could be parsed. I feel this would enhance the yaml files considerably.
> 
> I did precisely that at various points for parsing MM scenario files.
> 
> Admittedly, that would require parsing the JsonNode directly rather than with annotations only. But my feeling is that this should be doable by changing only how SourceableValue is parsed, meaning at only one place.

I appreciate that the `value:` fields are not great for readability, but parsing the nodes directly rather than using annotation would involve a lot more code and work on the Java side both now and in the future. Right now the yaml structure is set up to just work with the Java pretty much out of the box which I find highly appealing. 

The ultimate goal is to build some additional tools shortly that will allow for editing the planetary files, so that people aren't regularly rooting around in them by hand.

AaronGullickson commented:
Ok, @SJuliez. Trying to think through how I might be able to do what you are asking with a custom deserializer and I am getting stuck on the issue that `SourceableValue` is a generic class. This is what I have so far:

```java
    public SourceableValue deserialize(final JsonParser jsonParser, final DeserializationContext context) {
            try {
                JsonNode node = jsonParser.getCodec().readTree(jsonParser);
                JsonNode nodeValue = node.get("value");
                String source = null;
                Object obj;
                if(null == nodeValue) {
                    // then we just have the value here
                    //TODO: process the node as an object of its appropriate type
                } else {
                    // then we have a source: value: pair
                    JsonNode nodeSource = node.get("source");
                    if(null != nodeSource) {
                        source = nodeSource.asText();
                    }
                    //TODO: process the nodeValue as an object of the appropriate type
                }
                //SourceableValue<obj.class> sourceableValue = new SourceableValue<obj.class>();
                //sourceableValue.source = source;
                //sourceableValue.value = obj;
                return sourceableValue;
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }
```

I can use information about the node to figure out whether to process the current node or its child node of value, but I can't quite figure out how I would process the actual value field, because this is a generic class where the `value` field can refer to a variety of objects.

SJuliez commented:
OK cool! I suppose the class cannot be determined with certainty from the yaml alone, especially not stuff that uses enums or something. This means that the code must have a method to retrieve the class from the name of the field, 
public Class<?> getSourceableValueClass(String fieldName) { if (fieldName.equals("radius")) return Double.class; ... }
Then the deserialize could instantiate the value, hopefully (?)

(better return switch (fieldName) "radius", "distance" -> Double.class; etc. I suppose)

AaronGullickson commented:
@rjhancock @SJuliez Could you please look at commit 0d27a1a5e3 and give me your thoughts. I added a custom deserializer to `SourceableValue` that does what @SJuliez was asking for, which I think will lead to better yaml. I ran it on my test cases and its working fine. I will need to re-run the data conversion process to simplify the yaml structure and that takes awhile, so I probably won't be ready to merge until tomorrow, but if you guys can both sign off on the additional code that would be great.

rjhancock commented:
> OK cool! I suppose the class cannot be determined with certainty from the yaml alone

Correct. The way the YAMLFactory, at least in appearance, works is it just outputs/reads the values of the variables within the class and we are the ones that have to tell it what class to load it into. However, it will figure out if something is a child object and act accordingly as well.

The changes I'm making to the ImageAtlas require no annotations and just works. It saves all the data, and appears to read it back in just fine. But it's not a complicated set of objects either.

SJuliez commented:
This looks very good to me. Do the values end up being all SourceableValue<Object> or do they transform themselves into e.g. SourceableValue<Double>?
